packages/principal/src/bls/signer.js

```diff
@@ -0,0 +1,243 @@
+import * as Signer from '../signer.js'
+import { varint } from 'multiformats'
+import { base64pad } from 'multiformats/bases/base64'
+import { base16 } from 'multiformats/bases/base16'
+
+/** @type {API.MulticodecCode<0x1309, 'bls12-381-private-key'>} */
+
+export const code = 0x1309
+export const name = Verifier.name
+
+const PREFIX_SIZE = varint.encodingLength(code)
+export const PREFIX = varint.encodeTo(code, new Uint8Array(PREFIX_SIZE))
+
+export const SIZE = 32
+
+export const { signatureCode, signatureAlgorithm } = Verifier
+
+export const generate = async () => derive(bls.utils.randomPrivateKey())
+
+/**
+ * Derives issuer from 32 byte long secret key.
+ *
+ * @param {Uint8Array} secret
+ * @returns {API.BLSSigner}
+ */
+const derive = secret => {
+  if (secret.byteLength !== SIZE) {
+    throw new Error(
+      `Expected Uint8Array with byteLength of ${SIZE} instead not ${secret.byteLength}`
+    )
+  }
+
+  const bytes = new Uint8Array(SIZE + PREFIX_SIZE)
+
+  bytes.set(PREFIX, 0)
+  bytes.set(secret, PREFIX_SIZE)
+
+  return new BLSSigner(bytes)
+}
+
+/**
+ * @param {API.SignerArchive<API.DID, typeof signatureCode>} archive
+ * @returns {API.BLSSigner}
+ */
+export const from = ({ id, keys }) => {
+  if (id.startsWith('did:key:')) {
+    const key = keys[/** @type {API.DIDKey} */ (id)]
+    if (key instanceof Uint8Array) {
+      return decode(key)
+    }
+  }
+  throw new TypeError(`Unsupported archive format`)
+}
+
+/**
+ * @param {Uint8Array} bytes
+ * @returns {API.BLSSigner}
+ */
+export const decode = bytes => {
+  if (bytes.byteLength !== SIZE + PREFIX_SIZE) {
+    throw new Error(
+      `Expected Uint8Array with byteLength of ${
+        SIZE + PREFIX_SIZE
+      } instead not ${bytes.byteLength}`
+    )
+  }
+
+  {
+    const [keyCode] = varint.decode(bytes)
+    if (keyCode !== code) {
+      throw new Error(`Given bytes must be a multiformat with ${code} tag`)
+    }
+  }
+
+  return new BLSSigner(bytes)
+}
+
+/**
+ * @param {API.BLSSigner} signer
+ * @return {API.ByteView<API.BLSSigner & CryptoKeyPair>}
+ */
+export const encode = signer => {
+  const out = signer.encode()
+
+  return out
+}
+
+/**
+ * @template {string} Prefix
+ * @param {API.BLSSigner} signer
+ * @param {API.MultibaseEncoder<Prefix>} [encoder]
+ */
+export const format = (signer, encoder) =>
+  (encoder || base64pad).encode(encode(signer))
+
+/**
+ * @template {string} Prefix
+ * @param {string} principal
+ * @param {API.MultibaseDecoder<Prefix>} [decoder]
+ * @returns {API.BLSSigner}
+ */
+export const parse = (principal, decoder) =>
+  decode((decoder || base64pad).decode(principal))
+
+/**
+ *
+ * @param {string} archive
+ */
+export const fromFilecoinWallet = archive => {
+  const bytes = base16.baseDecode(archive)
+  const data = JSON.parse(new TextDecoder().decode(bytes))
+  if (data.Type !== 'bls') {
+    throw new TypeError(`Unsupported key type ${data.Type}`)
+  }
+
+  // For whatever reason bytes order is reversed
```

Comment:
Is there any point in a spec or whatever related to this?

Reply:
I have not found anything and spend hours trying to figure out why things were incompatible until I have came across this https://github.com/drand/tlock-js/blob/410e80691e236f25318749122a7ab6e6f6e2d8dc/src/crypto/utils.ts#L37-L38

My guess is that's how go serializes them, but I have not managed to find actual code that does it to check that assumbtion.